### PR TITLE
Package pytest

### PR DESCRIPTION
**Pull request by rth**
This packages pytest as discussed in https://github.com/iodide-project/pyodide/issues/69.

Also added a test to check that all built packages can be imported (unless it was already present somewhere and I missed it).

**Update:** this PR aims to be an example of adaptation of `meta.yaml` from conda-forge, following points would need to be addressed,

 - [x] support for sha256 checksum (used in conda forge) https://github.com/iodide-project/pyodide/pull/103
 - [x] ~~script to automatically convert `meta.yaml` from conda-forge to `pyodide` format~~ (will do in some unrelated PR)

### CODE REVIEW

**Comment by rth** [comment]
Currently `import pytest` fails, as illstrated in the following test,
```
$ pytest test/test_common.py::"test_package_imports[firefox-pytest]" -x -v
[..]
___________________________________________________________ test_package_imports[firefox-pytest] ____________________________________________________________


selenium = <test.conftest.FirefoxWrapper object at 0x7f4ae5f53320>, name = 'pytest'

    @pytest.mark.parametrize('name', registered_packages())
    def test_package_imports(selenium, name):
        selenium.load_package(name)
>       selenium.run('import %s' % name)

test/test_common.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/conftest.py:53: in run
    'return pyodide.runPython({!r})'.format(code))
test/conftest.py:60: in run_js
    return self.driver.execute_script(catch)
../../.miniconda3/envs/pyodide-env/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py:635: in execute_script
    'args': converted_args})['value']
../../.miniconda3/envs/pyodide-env/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py:320: in execute
    self.error_handler.check_response(response)

[..]
>       raise exception_class(message, screen, stacktrace)
E       selenium.common.exceptions.JavascriptException: Message: Error: Traceback (most recent call last):
E         File "/lib/python3.6/site-packages/pyodide.py", line 33, in eval_code
E           exec(compile(mod, '<exec>', mode='exec'), ns, ns)
E         File "<exec>", line 1, in <module>
E         File "/lib/python3.6/site-packages/pytest.py", line 9, in <module>
E           from _pytest.config import main, UsageError, cmdline, hookspec, hookimpl
E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 10, in <module>
E           import py
E       ModuleNotFoundError: No module named 'py'

../../.miniconda3/envs/pyodide-env/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py:242: JavascriptException
```
so I guess I will also need to package https://github.com/pytest-dev/py.

What I'm not sure about, is what happens when a package has dependencies. For instance, `pytest` depends on `pluggy`, `atomicwrites` and `more-itertools` among others. When first running `make -C pakages`, to build pytest, I see lines of the form,
```
Searching for atomicwrites>=1.0
Reading https://pypi.python.org/simple/atomicwrites/
Downloading https://files.pythonhosted.org/packages/a1/e1/2d9bc76838e6e6667fde5814aa25d7feb93d6fa471bf6816daac2596e8b2/atomicwrites-1.1.5.tar.gz#sha256=240831ea22da9ab882b551b31d4225591e5e447a68c5e188db5b89ca1d487585
Best match: atomicwrites 1.1.5
Processing atomicwrites-1.1.5.tar.gz
Writing /tmp/easy_install-fd72wtma/atomicwrites-1.1.5/setup.cfg
Running atomicwrites-1.1.5/setup.py -q bdist_egg --dist-dir /tmp/easy_install-fd72wtma/atomicwrites-1.1.5/egg-dist-tmp-os1io1v0
no previously-included directories found matching 'docs/_build'
zip_safe flag not set; analyzing archive contents...
Moving atomicwrites-1.1.5-py3.6.egg to /home/rth/src/pyodide/cpython/build/3.6.4/host/lib/python3.6/site-packages
Adding atomicwrites 1.1.5 to easy-install.pth file
```
so ths mean that it installs e.g. `atomicwrites` for CPython but that import of `atomicwrites` will not work from selenium unless a specific package is added in pyiodide/packages, right? Meaning that all of them would need to be packaged?

**Comment by mdboom** [comment]
That is correct: all of the dependencies would need to be packaged, and have their names added to the `requirements/run` section of py.test's `meta.yaml`.  Automatically extracting the dependencies from a pip package is basically an unsolvable problem -- even `conda-build` punted on it and requires that they are listed explicitly in the `meta.yaml`

File: test/test_common.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    pkg_dir = os.path.join(base_dir, '..', 'packages')
+    packages = [name for name in os.listdir(pkg_dir)
+                if os.path.isdir(os.path.join(pkg_dir, name))]
+    return packages
+
+
+@pytest.mark.parametrize('name', registered_packages())
+def test_package_imports(selenium, name):
+    selenium.load_package(name)
+    selenium.run('import %s' % name)
```
Comment (mdboom) [inline test/test_common.py]: This is a nice test to have, but it won't work in all cases.  The package name and the Python module path aren't always the same, see `python-dateutil`/`dateutil`.  What conda does is it allows the `meta.yaml` to specify a smoke test, e.g.:

```
test:
  imports:
    - dateutil
```

**Comment by mdboom** [comment]
This reminds me: I had an idea a while ago to make a tool that would create a `meta.yaml` from a PyPI entry, and prepopulate it with the latest version, the url and sha256 hash etc.  You'd still have to manually set up the requirements, but it would go a long way to simplifying the creation of pure Python packages.

File: test/test_common.py (hunk 1 above)
Comment (rth) [reply test/test_common.py]: You are right, will remove it form this PR and added a separate PR to parse the `test` section of `meta.yaml` in https://github.com/iodide-project/pyodide/pull/98

**Comment by rth** [comment]
I think I'll use this PR as a use case of adapting conda-forge feedstocks to pyodide as pytest [has a number of dependencies](https://github.com/conda-forge/pytest-feedstock/blob/master/recipe/meta.yaml) so some form of semi-automatic `meta.yaml` conversion would help.

**Comment by rth** [comment]
Trying to close this (will look into conda-forge feedstock adaption separately).

The current status is that all pytest dependencies are packaged however there are still the following issues,
 - `six` fails to load at `selenium.load_package` with a timeout without any additional information. I am not quite sure how to debug it.
 - actually running pytest on some package e.g. numpy with,
   ```py
    selenium.load_package('pytest')
    selenium.run("import pytest;"
                 "pytest.main(['--collect-only', '--pyargs', 'numpy'])")
   ```
   yields a `PermissionError`  in `os.dup2`,
   <details>
   
   ```
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 64, in pytest_load_initial_conftests
    E           capman.start_global_capturing()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 102, in start_global_capturing
    E           self._global_capturing.start_capturing()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 391, in start_capturing
    E           self.out.start()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 485, in start
    E           os.dup2(self.tmpfile_fd, self.targetfd)
    E       PermissionError: [Errno 13] Permission denied
    E       
    E       During handling of the above exception, another exception occurred:
    E       
    E       Traceback (most recent call last):
    E         File "/lib/python3.6/site-packages/pyodide.py", line 35, in eval_code
    E           return eval(compile(expr, '<eval>', mode='eval'), ns, ns)
    E         File "<eval>", line 1, in <module>
    E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 55, in main
    E           config = _prepareconfig(args, plugins)
    E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 183, in _prepareconfig
    E           config._ensure_unconfigure()
    E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 597, in _ensure_unconfigure
    E           fin()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 106, in stop_global_capturing
    E           self._global_capturing.pop_outerr_to_orig()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 397, in pop_outerr_to_orig
    E           out, err = self.readouterr()
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 437, in readouterr
    E           self.out.snap() if self.out is not None else "",
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 526, in snap
    E           res = FDCaptureBinary.snap(self)
    E         File "/lib/python3.6/site-packages/_pytest/capture.py", line 492, in snap
    E           self.tmpfile.truncate()
    E       PermissionError: [Errno 13] Permission denied

    ../../.miniconda3/envs/pyodide-env/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py:242: JavascriptException
   ```
   </details>
   so I imagine more patching is needed to make sure pytest doesn't use unsupported functionality..

**Comment by mdboom** [comment]
Sorry this isn't documented: `six` is included in the main filesystem file in `pyodide.asm.data`.  So you can think of it as "built-in".  Since virtually every significant package uses it, and it's rather small, I thought I'd avoid one round-trip by always including it.  I think just removing it as a package here should be fine.

It looks like that use of `dup2` is so that py.test can capture all stdout/stderr at the os level and redirect it to a temporary file.  *Perhaps* (and I haven't confirmed this) emscripten/musl can't do that.  It doesn't seem that the permission error is related to creation of the tempfile -- the `/tmp` directory is there and it's writable by an ordinary user (I'm actually not sure if the emscripten virtual filesystem enforces permissions -- again, haven't confirmed that).  I'm going to keep investigating for a bit.

**Comment by rth** [comment]
Thanks for the very helpful pointers! Removed six and rebased. Will try to play more with pytest tomorrow..

**Comment by mdboom** [comment]
I think this helps pytest.  emscripten's "system library" doesn't seem to like duping unnamed files.  I don't know how much work it would be to fix that, but in the meantime, we could apply this patch at our end:

```
--- src/_pytest/capture.py      2018-08-22 14:57:33.391502645 -0400
+++ build/lib/_pytest/capture.py        2018-08-22 14:57:33.397502659 -0400
@@ -10,7 +10,7 @@
 import os
 import io
 from io import UnsupportedOperation
-from tempfile import TemporaryFile
+from tempfile import NamedTemporaryFile as TemporaryFile
 
 import six
 import pytest
```

**Comment by rth** [comment]
Thanks ! Now pytest collection appears to work (or at least it doesn't fail). 

Though I have had a hard time finding a package that bundles test files to try this on: numpy doesn't https://github.com/iodide-project/pyodide/issues/126. pandas does but in 0.22.0 there is hard tests dependency in  a `conftest.py` on moto (https://github.com/pandas-dev/pandas/issues/20619), so all tests cannot be run. 

Running pytest on `Path(pandas.__file__).parent / "tests" / "frame" / "test_sorting.py"` gives another `OSError: [Errno 9] Bad file descriptor` somewhere in `py.io.TerminalWriter.write`, 

<details>

```
E       selenium.common.exceptions.JavascriptException: Message: Error: Traceback (most recent call last):
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 178, in wrap_session
E           session.exitstatus = doit(config, session) or 0
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 215, in _main
E           config.hook.pytest_runtestloop(session=session)
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
E           return outcome.get_result()
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
E           raise ex[1].with_traceback(ex[2])
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
E           res = hook_impl.function(*args)
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 236, in pytest_runtestloop
E           item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
E           return outcome.get_result()
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
E           raise ex[1].with_traceback(ex[2])
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
E           res = hook_impl.function(*args)
E         File "/lib/python3.6/site-packages/_pytest/runner.py", line 64, in pytest_runtest_protocol
E           runtestprotocol(item, nextitem=nextitem)
E         File "/lib/python3.6/site-packages/_pytest/runner.py", line 73, in runtestprotocol
E           rep = call_and_report(item, "setup", log)
E         File "/lib/python3.6/site-packages/_pytest/runner.py", line 163, in call_and_report
E           hook.pytest_runtest_logreport(report=report)
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
E           return outcome.get_result()
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
E           raise ex[1].with_traceback(ex[2])
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
E           res = hook_impl.function(*args)
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 374, in pytest_runtest_logreport
E           self.write_fspath_result(rep.nodeid, letter)
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 272, in write_fspath_result
E           self._tw.write(res)
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 224, in write
E           write_out(self._file, markupmsg)
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 384, in write_out
E           fil.flush()
E       OSError: [Errno 9] Bad file descriptor
E       
E       During handling of the above exception, another exception occurred:
E       
E       Traceback (most recent call last):
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 191, in wrap_session
E           config.notify_exception(excinfo, config.option)
E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 623, in notify_exception
E           res = self.hook.pytest_internalerror(excrepr=excrepr, excinfo=excinfo)
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
E           return outcome.get_result()
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
E           raise ex[1].with_traceback(ex[2])
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
E           res = hook_impl.function(*args)
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 327, in pytest_internalerror
E           self.write_line("INTERNALERROR> " + line)
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 294, in write_line
E           self.ensure_newline()
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 285, in ensure_newline
E           self._tw.line()
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 236, in line
E           self.write('\n')
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 224, in write
E           write_out(self._file, markupmsg)
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 369, in write_out
E           fil.write(msg)
E       OSError: [Errno 9] Bad file descriptor
E       
E       During handling of the above exception, another exception occurred:
E       
E       Traceback (most recent call last):
E         File "/lib/python3.6/site-packages/pyodide.py", line 35, in eval_code
E           return eval(compile(expr, '<eval>', mode='eval'), ns, ns)
E         File "<eval>", line 1, in <module>
E         File "/lib/python3.6/site-packages/_pytest/config/__init__.py", line 64, in main
E           return config.hook.pytest_cmdline_main(config=config)
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
E           return outcome.get_result()
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
E           raise ex[1].with_traceback(ex[2])
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
E           res = hook_impl.function(*args)
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 208, in pytest_cmdline_main
E           return wrap_session(config, _main)
E         File "/lib/python3.6/site-packages/_pytest/main.py", line 201, in wrap_session
E           session=session, exitstatus=session.exitstatus
E         File "/lib/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
E           firstresult=hook.spec_opts.get('firstresult'),
E         File "/lib/python3.6/site-packages/pluggy/callers.py", line 196, in _multicall
E           gen.send(outcome)
E         File "/lib/python3.6/site-packages/_pytest/terminal.py", line 584, in pytest_sessionfinish
E           self._tw.line("")
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 236, in line
E           self.write('\n')
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 224, in write
E           write_out(self._file, markupmsg)
E         File "/lib/python3.6/site-packages/py/_io/terminalwriter.py", line 369, in write_out
E           fil.write(msg)
E       OSError: [Errno 9] Bad file descriptor

../../.miniconda3/envs/pyodide-env/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py:242: JavascriptException
````
</details>

Maybe I should read more about the OS capabilities of emscripten, in terms of what is allowed when interacting with file handlers and what isn't. Also I'm not familiar with the pytest code base, maybe there is a way of disabling part of this stream manipulation without hurting the core functionality too much.. Will need to think more about this..

**Comment by mdboom** [comment]
> Maybe I should read more about the OS capabilities of emscripten, in terms of what is allowed when interacting with file handlers and what isn't. 

There not a [ton of docs on this](https://kripken.github.io/emscripten-site/docs/api_reference/Filesystem-API.html).  I generally end up reading `library_syscalls.js` in emscripten when I want to see how a particular system call is implemented.

**Comment by rth** [comment]
I think this PR is ready for another review.

In pytest ended up replacing all `os.dup2(fh1, fh2)` with  `fh2 = os.dup(fh1)`, which is probably not ideal, but might be good enough for the first try. For instance, running `test/test_testing.py::test_pytest` with `-s` to print the stdout yields, 
```
% pytest test/test_testing.py -v -s  -k chrome
Spawning webserver...
Running webserver...
serving at port 39591
==================================================================== test session starts ====================================================================
platform linux -- Python 3.6.6, pytest-3.7.1, py-1.5.4, pluggy-0.7.1 -- /home/rth/.miniconda3/envs/pyodide-env/bin/python
cachedir: .pytest_cache
rootdir: /home/rth/src/pyodide, inifile: setup.cfg
plugins: instafail-0.4.0
collected 2 items / 1 deselected                                                                                                                            

test/test_testing.py::test_pytest[chrome] PASSED/lib/python3.6/site-packages/numpy/core/tests
============================= test session starts ==============================
platform emscripten -- Python 3.6.4, pytest-3.6.3, py-1.5.4, pluggy-0.7.1
rootdir: /lib/python3.6/site-packages/numpy/core, inifile:
collected 9 items                                                              

lib/python3.6/site-packages/numpy/core/tests/test_api.py .........       [100%]

=========================== 9 passed in 3.73 seconds ===========================


========================================================== 1 passed, 1 deselected in 22.82 seconds ==========================================================
```

which currently runs numpy tests in `numpy/core/tests/test_api.py`. There is probably more ways to make this better, but it's a start.

Also packaged nose as that was necessary to run numpy tests.

**Comment by rth** [comment]
Though numerous `os.dup` call make us run out of available file handlers after some number of tests. For instance,  when running the full numpy tests suite, we end up with,
```py
  File "/lib/python3.6/pathlib.py", line 1161, in open
  File "/lib/python3.6/pathlib.py", line 1015, in _opener
  File "/lib/python3.6/pathlib.py", line 387, in wrapped
OSError: [Errno 24] No file descriptors available: '/lib/python3.6/site-packages/numpy/core/.pytest_cache/v/cache/nodeids'
```
so in the latest version here I am removing all `os.dup2` and it seems works well enough (I think it was used for capturing stdout which is not critical)..

**Comment by mdboom** [comment]
Yes, I agree that capturing stdout is a nice-to-have, but not critical.

**Comment by rth** [comment]
Do you have any other comments about this PR @mdboom ? This adds a few packages, but they are pure Python and don't take too long to build. Also having pytest is very useful for running any kind of tests.

**Comment by mdboom** [comment]
Looks good to me.  I was only holding off merging because I didn't notice you took the "WIP" off the title...  :)

**Pull request merged**